[CSS Grid Layout] Fix grid-{row, column, area} shorthand CSSOM serialization

zakariaridouh opened the pull request:
#### 3725a9f39e9d4973d8090c6dc135eafb69b81a76
<pre>
[CSS Grid Layout] Fix grid-{row, column, area} shorthand CSSOM serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=149890">https://bugs.webkit.org/show_bug.cgi?id=149890</a>
&lt;<a href="https://rdar.apple.com/132530604">rdar://132530604</a>&gt;

Reviewed by Matthieu Dubet.

Fix grid-area serialization, which previously was not fully following
the CSS Spec. Gains us a small amount on interop, and we are now passing
all tests in the grid-area-valid test[1].

This was done by adding a new child class of `CSSValue` called
`CSSGridLineValue`, and using it instead of `CSSValueList`, and
implementing the `customCSSText()` function in line with the spec.

[1] <a href="https://wpt.live/css/css-grid/parsing/grid-area-valid.html">https://wpt.live/css/css-grid/parsing/grid-area-valid.html</a>

    * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-area-shorthand-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-area-valid-expected.txt:
    * Source/WebCore/Sources.txt:
    * Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    * Source/WebCore/css/CSSGridLineNamesValue.cpp:
    * Source/WebCore/css/CSSGridLineValue.cpp: Added.
    (WebCore::CSSGridLineValue::customCSSText const):
    (WebCore::CSSGridLineValue::CSSGridLineValue):
    (WebCore::CSSGridLineValue::create):
    (WebCore::CSSGridLineValue::equals const):
    * Source/WebCore/css/CSSGridLineValue.h: Added.
    * Source/WebCore/css/CSSTimingFunctionValue.h:
    * Source/WebCore/css/CSSValue.cpp:
    (WebCore::CSSValue::visitDerived):
    * Source/WebCore/css/CSSValue.h:
    (WebCore::CSSValue::isGridLineValue const):
    * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
    (WebCore::CSSPropertyParserHelpers::consumeGridLine):
    * Source/WebCore/style/StyleBuilderConverter.h:
    (WebCore::Style::BuilderConverter::createGridPosition):

Canonical link: <a href="https://commits.webkit.org/283234@main">https://commits.webkit.org/283234@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d804e4112e1ee8b0a970622a9608b05b8f3a688c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17968 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69378 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15960 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67472 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52596 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16242 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52478 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11034 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33102 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38080 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13939 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14837 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59915 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71083 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9306 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13808 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59804 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9338 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56626 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60080 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/ignored-objects, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/list-markers (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14467 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7684 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1341 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40533 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42791 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41354 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/CSSGridLineValue.cpp:
```diff
@@ -0,0 +1,69 @@
+    Vector<String> parts;
+    StringBuilder builder;
+    if (m_spanValue)
+        parts.append(m_spanValue->cssText());
```
Why are we not using the builder all the time ?? We should get rid of the `Vector<String> parts`

zakariaridouh replied:
Was using `Vector<String> parts` to be able to pass it into `interleave` from `wtf/StringConcat.h` to use a separator between the values, as we can have 1, 2, or 3 values, while keeping the code clean (not having to use a flag to know if we want to append a separator)

zakariaridouh replied:
I don't see a `StringBuiler` function that lets us `toString()` with a specific separator. Are you aware of one? 

Because otherwise it would be best to just use a container like `Vector<String>` then concat with separator similar to how we are currently doing.

mdubet commented on Source/WebCore/css/CSSGridLineValue.h:
```diff
@@ -0,0 +1,48 @@
+
+class CSSGridLineValue final : public CSSValue {
+public:
+    static Ref<CSSGridLineValue> create(RefPtr<CSSPrimitiveValue>, RefPtr<CSSPrimitiveValue>, RefPtr<CSSPrimitiveValue>);
```
We should use a moved RefPtr (`RefPtr&&`) for all of this to avoid the refcounter churn

mdubet commented on Source/WebCore/css/CSSGridLineValue.h:
```diff
@@ -0,0 +1,48 @@
+    static Ref<CSSGridLineValue> create(RefPtr<CSSPrimitiveValue>, RefPtr<CSSPrimitiveValue>, RefPtr<CSSPrimitiveValue>);
+
+    String customCSSText() const;
+    bool equals(const CSSGridLineValue& other) const { return customCSSText() == other.customCSSText(); }
```
Do we really want to check equality by going through serialisation ? Why are we not comparing the actual members ?

zakariaridouh replied:
When would the serialized values not be the same? Would we have a situation where the objects are equal but their serialized values are not? 

Regardless, I've switched it to comparing each individual value. =)

mdubet commented on Source/WebCore/css/CSSGridLineValue.cpp:
```diff
@@ -0,0 +1,69 @@
+        parts.append(m_gridLineName->cssText());
+
+    builder.append(
+        interleave(parts, [](const String& part) { return part; }, " "_s)
```
Do we want `WTFString::makeStringByJoining()` ?

zakariaridouh replied:
Sure, we can do that instead of manually calling StringBuilder.

mdubet commented on Source/WebCore/css/CSSGridLineValue.cpp:
```diff
@@ -0,0 +1,87 @@
+bool CSSGridLineValue::equals(const CSSGridLineValue& other) const
+{
+    if (m_spanValue) {
+        if (!other.m_spanValue || m_spanValue->cssText() != other.m_spanValue->cssText())
```
We should just recursively call `CSSPrimitiveValue::equals()` on the 3 members, not go through serialisation

mdubet commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1309,33 +1310,17 @@ inline bool BuilderConverter::createGridPosition(const CSSValue& value, GridPosi
-        currentValue = it != values.end() ? &downcast<CSSPrimitiveValue>(*it) : nullptr;
-    }
+    auto& gridLineValue = downcast<CSSGridLineValue>(value);
+    CSSPrimitiveValue* uncheckedSpanValue = gridLineValue.spanValue();
```
We should not have unsafe raw pointer on the stack. They should be stored in a `RefPtr`.

mdubet reviewed (changes requested):
I've added some comments

mdubet commented on Source/WebCore/css/CSSGridLineValue.h:
```diff
@@ -0,0 +1,52 @@
+public:
+    static Ref<CSSGridLineValue> create(RefPtr<CSSPrimitiveValue>&&, RefPtr<CSSPrimitiveValue>&&, RefPtr<CSSPrimitiveValue>&&);
+
+    String customCSSText() const;
```
`override` ?

zakariaridouh replied:
Both `cssText()` and `customCSSText()` are not virtual function in `CSSValue`, so I believe we don't need to add `override` in this situation.

mdubet replied:
Oh right we do the dispatching manually with `visitDerived` 😅 Nevermind

zakariaridouh replied:
Any good reason for doing that? Why not have the function be virtual on CSSValue and force implementation in child subclasses?

mdubet commented on Source/WebCore/css/CSSGridLineValue.h:
```diff
@@ -0,0 +1,52 @@
+    static Ref<CSSGridLineValue> create(RefPtr<CSSPrimitiveValue>&&, RefPtr<CSSPrimitiveValue>&&, RefPtr<CSSPrimitiveValue>&&);
+
+    String customCSSText() const;
+    bool equals(const CSSGridLineValue& other) const;
```
Ditto

zakariaridouh replied:
Ditto

zakariaridouh committed:
[CSS Grid Layout] Fix grid-{row, column, area} shorthand CSSOM serialization
https://bugs.webkit.org/show_bug.cgi?id=149890
<rdar://132530604>

Reviewed by Matthieu Dubet.

Fix grid-area serialization, which previously was not fully following
the CSS Spec. Gains us a small amount on interop, and we are now passing
all tests in the grid-area-valid test[1].

This was done by adding a new child class of `CSSValue` called
`CSSGridLineValue`, and using it instead of `CSSValueList`, and
implementing the `customCSSText()` function in line with the spec.

[1] https://wpt.live/css/css-grid/parsing/grid-area-valid.html

    * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-area-shorthand-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-area-valid-expected.txt:
    * Source/WebCore/Sources.txt:
    * Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    * Source/WebCore/css/CSSGridLineNamesValue.cpp:
    * Source/WebCore/css/CSSGridLineValue.cpp: Added.
    (WebCore::CSSGridLineValue::customCSSText const):
    (WebCore::CSSGridLineValue::CSSGridLineValue):
    (WebCore::CSSGridLineValue::create):
    (WebCore::CSSGridLineValue::equals const):
    * Source/WebCore/css/CSSGridLineValue.h: Added.
    * Source/WebCore/css/CSSTimingFunctionValue.h:
    * Source/WebCore/css/CSSValue.cpp:
    (WebCore::CSSValue::visitDerived):
    * Source/WebCore/css/CSSValue.h:
    (WebCore::CSSValue::isGridLineValue const):
    * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
    (WebCore::CSSPropertyParserHelpers::consumeGridLine):
    * Source/WebCore/style/StyleBuilderConverter.h:
    (WebCore::Style::BuilderConverter::createGridPosition):

Canonical link: https://commits.webkit.org/283234@main
